lte/cloud/go/services/subscriberdb/digest.go

```diff
@@ -0,0 +1,121 @@
+limitations under the License.
+*/
+
+package subscriberdb
+
+import (
+	"magma/lte/cloud/go/lte"
+	"magma/lte/cloud/go/serdes"
+	lte_models "magma/lte/cloud/go/services/lte/obsidian/models"
+	"magma/orc8r/cloud/go/mproto"
+	mproto_protos "magma/orc8r/cloud/go/mproto/protos"
+	"magma/orc8r/cloud/go/services/configurator"
+
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
+)
+
+const defaultSubProfile = "default"
+
+// GetDigest returns a deterministic digest of the current configurations of a
+// network, which is a concatenation of its subscribers digest and apn resources digest.
+func GetDigest(networkID string) (string, error) {
+	// HACK: Workaround to decouple apn resources data from subscriber data
+	// despite current construction logic.
+	//
+	// The subscribers and apn resources digests are separately generated so
+	// that the digests are per-network, instead of having to account for
+	// gateway-specific apn resources configs for subscribers. We can do this
+	// because APN resource IDs are unique within each network.
+	subscribersDigest, err := getSubscribersDigest(networkID)
+	if err != nil {
+		return "", err
+	}
+
+	apnResourcesDigest, err := getApnResourcesDigest(networkID)
+	if err != nil {
+		return "", err
+	}
+
+	digest := subscribersDigest + apnResourcesDigest
+	return digest, nil
+}
+
+// getSubscribersDigest returns a deterministic digest of all subscribers in the network.
+func getSubscribersDigest(networkID string) (string, error) {
+	apnsByName, err := LoadApnsByName(networkID)
+	if err != nil {
+		return "", err
+	}
+
+	pageDigestsByToken := map[string][]byte{}
+	token := ""
+	curPage := 0
+	foundEmptyToken := false
+
+	for !foundEmptyToken {
+		subProtosById := map[string]proto.Message{}
+		subProtos, nextToken, err := LoadSubProtosPage(0, token, networkID, apnsByName, lte_models.ApnResources{})
+		if err != nil {
+			return "", err
+		}
+		for _, subProto := range subProtos {
+			subProtosById[subProto.Sid.Id] = subProto
+		}
+
+		// Take a digest per page to be combined in the end, to avoid saving all
+		// subscriber data in the memory at once
+		pageDigest, err := mproto.HashManyDeterministic(subProtosById)
+		if err != nil {
+			return "", nil
+		}
+		pageDigestsByToken[string(curPage)] = []byte(pageDigest)
+
+		foundEmptyToken = nextToken == ""
+		token = nextToken
+		curPage++
+	}
+	digestProto := &mproto_protos.ProtosByID{BytesById: pageDigestsByToken}
+	return mproto.HashDeterministic(digestProto)
+}
+
+// getApnResourcesDigest returns a deterministic digest of the apn resources configurations
+// in a network.
+func getApnResourcesDigest(networkID string) (string, error) {
+	apnResourceEnts, _, err := configurator.LoadAllEntitiesOfType(
+		networkID, lte.APNResourceEntityType,
+		configurator.EntityLoadCriteria{LoadConfig: true},
+		serdes.Entity,
+	)
+
+	if err != nil {
+		return "", err
+	}
+	apnResourcesProtosByID := map[string]proto.Message{}
```

Comment:
Nit: group by "paragraph"
```suggestion
	)
	if err != nil {
		return "", err
	}
	
	apnResourcesProtosByID := map[string]proto.Message{}
```